packages/expect/src/toThrowMatchers.js

```diff
@@ -7,203 +7,225 @@
-            printActualErrorMessage(error),
-      pass,
-    };
-  } else {
-    throw new Error(
-      matcherErrorMessage(
-        matcherHint('[.not]' + matcherName, undefined, undefined),
-        `${EXPECTED_COLOR(
-          'expected',
-        )} value must be a string or regular expression or Error`,
-        printWithType('Expected', expected, printExpected),
-      ),
-    );
-  }
-};
+
+    if (error && !error.message && !error.name) {
+      error = new Error(error);
+    }
+
+    if (expected === undefined) {
+      return toThrow(matcherName, options, error);
+    } else if (typeof expected === 'function') {
+      return toThrowExpectedClass(matcherName, options, error, expected);
+    } else if (typeof expected === 'string') {
+      return toThrowExpectedString(matcherName, options, error, expected);
+    } else if (expected !== null && typeof expected.test === 'function') {
+      return toThrowExpectedRegExp(matcherName, options, error, expected);
+    } else if (expected !== null && typeof expected === 'object') {
+      return toThrowExpectedObject(matcherName, options, error, expected);
+    } else {
+      throw new Error(
+        matcherErrorMessage(
+          matcherHint(matcherName, undefined, undefined, options),
+          `${EXPECTED_COLOR(
+            'expected',
+          )} value must be a string or regular expression or class or error`,
+          printWithType('Expected', expected, printExpected),
+        ),
+      );
+    }
+  };
 
 const matchers: MatchersObject = {
-  toThrow: createMatcher('.toThrow'),
-  toThrowError: createMatcher('.toThrowError'),
+  toThrow: createMatcher('toThrow'),
+  toThrowError: createMatcher('toThrowError'),
 };
 
-const toThrowMatchingStringOrRegexp = (
-  name: string,
-  error: ?Error,
-  pattern: RegExp,
-  value: RegExp | string | Error,
+const toThrowExpectedRegExp = (
+  matcherName: string,
+  options: MatcherHintOptions,
+  error?: Error,
+  expected: RegExp,
 ) => {
-  if (error && !error.message && !error.name) {
-    error = new Error(error);
-  }
+  const pass = error !== undefined && expected.test(error.message);
 
-  const pass = !!(error && error.message.match(pattern));
   const message = pass
     ? () =>
-        matcherHint('.not' + name, 'function', getType(value)) +
+        matcherHint(matcherName, undefined, undefined, options) +
         '\n\n' +
-        `Expected the function not to throw an error matching:\n` +
-        `  ${printExpected(value)}\n` +
-        printActualErrorMessage(error)
+        `Expected error pattern: ${printExpected(expected)}\n` +
+        // Possible improvement also for toMatch inverse highlight matching substring.
+        // $FlowFixMe: Cannot get error.message because property message is missing in undefined
+        `Received error message: ${printReceived(error.message)}\n` +
+        // $FlowFixMe: Cannot get error.stack because property stack is missing in undefined
```

Comment:
This is async code and theoretically there is a chance that `error` is mutated somewhere in between the calls. Can we guard this the same as on the other branch with `error !== undefined`?

This comment applies to all `$FlowFixMe`s added in this file

Reply:
As usual, your review comments are very helpful. Instead of the following code:

```js
if (error && !error.message && !error.name) {
  error = new Error(error);
}
```

which has incorrect edge cases for falsey exception values like `throw 0`

I will rewrite as robust destructuring to provide `message` or `name` as arguments

Reply:
Do you agree if exception is **primitive value**, then assertion should **fail for expected class**?

Now the following assertion **passes**, because of `if` statement in preceding comment:

```js
expect(() => { throw true; }).toThrow(Error);
```

And for your enjoyment, here is the mother of all edge cases: `() => { throw undefined; }`